src/core/raster/qgsrasterdataprovider.cpp

```diff
@@ -642,6 +644,231 @@ void QgsRasterDataProvider::writeXml( QDomDocument &doc, QDomElement &parentElem
                                   QString::number( mMaxOversampling ) );
 }
 
+QgsRasterAttributeTable *QgsRasterDataProvider::attributeTable( int bandNumber ) const
+{
+  try
+  {
+    return mAttributeTables.at( bandNumber ).get();
+  }
+  catch ( std::out_of_range const & )
+  {
+    return nullptr;
+  }
+}
+
+void QgsRasterDataProvider::setAttributeTable( int bandNumber, QgsRasterAttributeTable *attributeTable )
+{
+  if ( attributeTable )
+  {
+    mAttributeTables[ bandNumber ] = std::unique_ptr<QgsRasterAttributeTable>( attributeTable );
+  }
+  else
+  {
+    removeAttributeTable( bandNumber );
+  }
+}
+
+void QgsRasterDataProvider::removeAttributeTable( int bandNumber )
+{
+  if ( mAttributeTables.find( bandNumber ) !=  std::end( mAttributeTables ) )
+  {
+    mAttributeTables.erase( bandNumber );
+  }
+}
+
+bool QgsRasterDataProvider::writeFileBasedAttributeTable( int bandNumber, const QString &path ) const
+{
+
+  const QgsRasterAttributeTable *rat { attributeTable( bandNumber ) };
+  if ( ! rat )
+  {
+    return false;
+  }
+
+  QString nPath { path };
+  if ( ! path.endsWith( QStringLiteral( ".vat" ) ) )
+  {
+    nPath += QStringLiteral( ".vat" );
+  }
+
+  QgsVectorFileWriter::SaveVectorOptions options;
+  options.driverName = QStringLiteral( "ESRI Shapefile" );
+  options.layerOptions = QStringList() << QStringLiteral( "SHPT=NULL" );
+
+  std::unique_ptr<QgsVectorFileWriter> writer { QgsVectorFileWriter::create( nPath, rat->qgisFields(), QgsWkbTypes::Unknown, QgsCoordinateReferenceSystem(), QgsCoordinateTransformContext(), options ) };
+  QgsFeatureList features { rat->qgisFeatures() };
+  writer->addFeatures( features );
+
+  return true;
+}
+
+bool QgsRasterDataProvider::readNativeAttributeTable()
+{
+  return false;
+}
+
+bool QgsRasterDataProvider::readFileBasedAttributeTable( int bandNumber, const QString &path )
+{
+  QgsVectorLayer ratLayer { path, QStringLiteral( "RAT layer" ), QStringLiteral( "ogr" ) };
```

Comment:
Again, could skip some overhead by creating an ogr provider directly here instead of a vector layer